Moored myth: Notifications

modernserf opened the pull request:
http://moored-myth.glitch.me/#StoryNotification

ehmorris commented:
A few visual things:

- I think notifications should have a max-width of something like 400px.
- Single-line notifications should't have more space below the text than above. I see in this case the dismiss button it creating the extra height.

modernserf commented:
- increased notification container width to 400px
- fixed uneven spacing

ehmorris commented on lib/notification.js:
```diff
@@ -194,7 +195,7 @@ const NotificationsContainer = styled.div`
   z-index: var(--z-notifications);
   top: 0;
   right: 0;
-  width: 18rem;
```
Ah sorry, I was looking at the Notification element itself and didn't notice the `Create Notification`, `Create Success Notification` and `Create Error Notification` buttons that display notifications within a container. If this is the only place where we intend to apply a width to the Notification components then `18rem` does look more correct than `25rem`.

modernserf replied:
👍

modernserf commented:
@clottman my understanding is that for text-only notifications you _dont_ want to capture focus, and use aria-live to notify users when a notification renders, but what should we do about notifications with buttons in them?

clottman commented:
@modernserf our pattern for notifications with buttons in them has been to strip out the non-text content in what we give to the aria-live section to read. So the live region should only announce 'Added cool-project to collection' even if the button also says 'take me there' with those words in a button.

clottman commented on lib/progress.js:
```diff
@@ -0,0 +1,63 @@
+      </div>
+      <div style={{ color: 'var(--colors-notice-text)', backgroundColor: 'var(--colors-notice-background)', padding: 'var(--space-1)' }}>
+        <p>Notification styles</p>
+        <Progress value={70} max={100}>
```
Voiceover is reading this as "indeterminate progress indicator" instead of "70% progress indicator" like it should be. I copied the markup (as well as some variations on the markup) and I hear the correct behavior in https://nebulous-crab.glitch.me/. 

Without doing more digging I'm kinda stumped on why this isn't working - I'm testing both in Chrome on Mac, in the same browser window.

clottman replied:
Found that someone else has experienced the same thing testing styled progress bars in Chrome with Voiceover. [see link](https://scottaohara.github.io/a11y_styled_form_controls/src/progress-bar/). This would explain why my demo app works - it's using unstyled progress bars. 

> Though the progress element is largely reporting the correct information to browsers, screen readers have quite a few kinks in how they announce the element to users.
> 
> At the time of testing, a styled progress bar won't be fully accessible in all screen reader and browser pairings. Instead it may be more appropriate to simply treat the progress bar as a visual decoration, hide it from screen readers, and provide visually hidden text as a means to consistently convey the information.

@modernserf what do you think about modifying the component to take this approach? Using the VisuallyHidden component for the text-only version and aria-hidden='true' for the `<progress>` element should accomplish what we need without much extra work/markup.

clottman commented on lib/notification.js:
```diff
@@ -0,0 +1,345 @@
+import { code, CodeExample, PropsDefinition, Prop } from './story-utils';
+
+const variants = {
+  notice: css`
```
Are these the same colors/font sizes we're using on the community site? The contrast is too low (green is 1.98, purple is 3:1, and red is 2.97. At this font size, the Accessibility Insights tool wants to see 4.5:1 contrast ratios. 

CC: @ttseng, @ehmorris

ttseng replied:
Yeah, there were some tweaks to the notification purple and error red we made a few months ago to improve contrast:

```
notice: #7460E1
error: #DC352C
```

to improve the contrast of `success` I'd suggest lightening the green and changing the text to black:

![image](https://user-images.githubusercontent.com/519336/62897234-d617d500-bd20-11e9-914b-5079b89567a3.png)

(the success green shown above is `#47F68D`

(already discussed with @ehmorris and we're both good with this change)

modernserf commented:
- updated the Progress component to use visually-hidden text + aria-hidden progress bar
- updated notification components to use the aria-live technique detailed in the linked article

clottman commented on lib/live-announcer.js:
```diff
@@ -0,0 +1,54 @@
+};
+
+const MessageBlock = ({ live, message = '' }) => (
+  <div role="log" aria-relevant="additions" aria-atomic="true" aria-live={live}>
```
Why not include `text` in `aria-relevant`? (Making it the default value, `additions text`). As is, if there are 3 messages in quick succession, it's possible the first one won't have gone to nothing yet and so the third one will be interpreted as a text change, not an addition.

clottman replied:
I _think_ proving this out, when I generated a bunch of notifications (some with the same text), only the unique text ones or the first ones of a certain text got announced. I could be wrong about what's causing that, though. 
![image](https://user-images.githubusercontent.com/4480480/62962752-2ba5bd80-bdc5-11e9-84e2-d57ade2a3027.png)

ehmorris commented:
I just realized that this is missing the onboarding notification type.

<img width="241" alt="Screen Shot 2019-08-13 at 1 40 58 PM" src="https://user-images.githubusercontent.com/654275/62963981-108c7b00-bdd0-11e9-8eb2-58494dd78ce7.png">

This uses the `#bfe3ff` color

sheridanvk commented on lib/notification.js:
```diff
@@ -0,0 +1,341 @@
+  right: 0;
+  width: 18rem;
+  padding: var(--space-1);
+`;
```
I agree with Edwin that it would be nice for our notifications to have a max width, and maybe even a standard width. Could we consider including this div by default in a Notification?

modernserf replied:
width and default close behavior are provided by NotificationContainer

modernserf replied:
see https://github.com/FogCreek/shared-components/pull/30#discussion_r312594833

sheridanvk replied:
Yeah I'm proposing that the top-level exported component be NotificationsContainer, with a pre-assigned width for consistency across community/editor, and the notification as children of the NotificationsContainer component. But this is ok for now, we can just make sure that we coordinate with community to ensure we're using the same widths.

modernserf replied:
NotificationsContainer is exported as part of NotificationsProvider

modernserf replied:
Do you _also_ want NotificationsContainer exported? I figure that notifications are mostly useful within a system -- the Notification component is designed to be used with 
`createNotification`, not so much as a free-standing component.

sheridanvk replied:
Hmm ok, I've been thinking about this slightly differently because we currently use Observables all over the place to control whether or not notifications show, which unless I'm mistaken wouldn't require the NotificationsProvider, but this is a good point at which to move away from that approach and use something more functional like createNotification. In which case this is probably fine as is.

sheridanvk replied:
I don't think I'm following how this works though. In the code (not the example in the documentation) I see that you created a component called Content, which contains the notifications and their means of being triggered but it's not clear to me what is being imported and what is being created by the end user. Could you change the documented example to show exactly how this NotificationsProvider component gets used in the context of a system, rather than the isolated snippets right now? We can talk through this on Monday as well if it would be easier.

modernserf replied:
`Notification`, `useNotifications` and `NotificationsProvider` are provided by shared components; the actual content of the notifications is done by the consumer. `Content` is a separate component because `useNotifications` consumes the context provided by `NotificationsProvider` and therefore must be used in a child component. 

In community, our root component looks something like

```js
const App = () => (
  <APIProvider>
    <CurrentUserProvider>
      <NotificationsProvider>
         {/* many layers */}
                          <AppContent />
         {/* many layers */}
      </NotificationsProvider>
    </CurrentUserProvider>
  </APIProvider>
)
```

And we have components defined for our common notifications, which look something like

```js
const AddedProjectToCollectionNotification = ({ project, collection, ...props }) => (
  <Notification 
    variant="success"  
    message={`Added ${project.domain} to collection ${collection.name}`}
    {...props}
  >
    <p >Added {project.domain} to collection {collection.name}</p>
    <Button as="a" variant="secondary" size="tiny" href={collection.fullUrl}>
      Take me there
    </Button>
  </Notification>
)
```

 Which we use in the app like

```js
const AddCollectionProjectPopover = ({ collection }) => {
  const api = useAPI();
  const { createNotification, createErrorNotification } = useNotifications();

  const addProjectToCollection = project => {
    api
      .addProjectToCollection({ project, collection })
      .then(
        () =>
          createNotification(props => (
            <AddedProjectToCollectionNotification
              project={project}
              collection={collection}
              {...props}
            />
          )),
        () => createErrorNotification("Could not add project to collection")
      );
  };

  // ...
};
```

modernserf commented on lib/notification.js:
```diff
@@ -0,0 +1,341 @@
+  & + & {
+    margin-top: var(--space-1);
+  }
+  &:hover ${CloseButton} {
```
clicking outside the notification to dismiss would lead to a lot of accidental dismissals

sheridanvk replied:
Possibly yeah, we can revisit after a bit of usage then :) how do you feel about making the close button visible by default?

clottman commented on lib/notification.js:
```diff
@@ -0,0 +1,343 @@
+      <Button
+        onClick={() =>
+          createNotification((props) => (
+            <Notification variant="onboarding" {...props} message="Create ana account to keep this project and edit it anywhere.">
```
missing a space in 'an a'

modernserf commented:
@clottman the live notification works more or less the same as react-aria-live; do we have problems with rapid notifications in Communnity?

clottman commented:
@modernserf we probably have the same issue there but haven't hit it because it doesn't happen much. I'd say don't worry about it for now.

clottman commented:
Sorry! Bad click.

sheridanvk commented:
Great, thank you - that's helpful. I am still confused though about your
earlier comment about "NotificationsContainer is exported with
NotificationProvider" - can you tell me how that currently works? I am
asking because I want to know how we're keeping the look of the
notification consistent between community and the editor, and I don't think
you're using this component as-is on the community side yet, right? I see
the notifications width currently being set in CSS bundled as part of the
CSS module (community site remix):
https://glitch.com/edit/#!/horse-chevre?path=src/components/notification/styles.styl:12:21

I tried implementing all this in
https://glitch.com/edit/#!/short-koala?path=app/app.jsx:30:65, but I'm
still running into issues (major error seems to be Uncaught (in promise)
TypeError: Cannot destructure property `createNotification` of 'undefined'
or 'null'.). Please can you point out what I'm doing wrong, or can we pair
on it later today?

On Sat, Aug 17, 2019 at 10:51 AM Justin Falcone <notifications@github.com>
wrote:

> *@modernserf* commented on this pull request.
> ------------------------------
>
> In lib/notification.js
> <https://github.com/FogCreek/shared-components/pull/30#discussion_r314949546>
> :
>
> > +  currentID += 1;
> +  return `${prefix}-${currentID}`;
> +};
> +
> +const NotificationContext = React.createContext();
> +export const NotificationsConsumer = NotificationContext.Consumer;
> +export const useNotifications = () => React.useContext(NotificationContext);
> +
> +const NotificationsContainer = styled.div`
> +  position: fixed;
> +  z-index: var(--z-notifications);
> +  top: 0;
> +  right: 0;
> +  width: 18rem;
> +  padding: var(--space-1);
> +`;
>
> Notification, useNotifications and NotificationsProvider are provided by
> shared components; the actual content of the notifications is done by the
> consumer. Content is a separate component because useNotifications
> consumes the context provided by NotificationsProvider and therefore must
> be used in a child component.
>
> In community, our root component looks something like
>
> const App = () => (
>   <APIProvider>
>     <CurrentUserProvider>
>       <NotificationsProvider>
>          {/* many layers */}
>                           <AppContent />
>          {/* many layers */}
>       </NotificationsProvider>
>     </CurrentUserProvider>
>   </APIProvider>
> )
>
> And we have components defined for our common notifications, which look
> something like
>
> const AddedProjectToCollectionNotification = ({ project, collection, ...props }) => (
>   <Notification
>     variant="success"
>     message={`Added ${project.domain} to collection ${collection.name}`}
>     {...props}
>   >
>     <p >Added {project.domain} to collection {collection.name}</p>
>     <Button as="a" variant="secondary" size="tiny" href={collection.fullUrl}>
>       Take me there
>     </Button>
>   </Notification>
> )
>
> Which we use in the app like
>
> const AddCollectionProjectPopover = ({ collection }) => {
>   const api = useAPI();
>   const { createNotification, createErrorNotification } = useNotifications();
>
>   const addProjectToCollection = project => {
>     api
>       .addProjectToCollection({ project, collection })
>       .then(
>         () =>
>           createNotification(props => (
>             <AddedProjectToCollectionNotification
>               project={project}
>               collection={collection}
>               {...props}
>             />
>           )),
>         () => createErrorNotification("Could not add project to collection")
>       );
>   };
>
>   // ...
> };
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/FogCreek/shared-components/pull/30?email_source=notifications&email_token=ACLFWLBJEYE33TR6DIJNGRTQFAF6RA5CNFSM4IKO3ZDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCB3RLPQ#discussion_r314949546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACLFWLC4BIV4RRHCGRWASO3QFAF6RANCNFSM4IKO3ZDA>
> .
>

modernserf commented:
@sheridan 
- NotificationsContainer is part of NotificationProvider
- ~community is not using shared-components; shared-components is _based on_ the existing stuff in community, but with a substantial number of changes

modernserf commented:
@sheridanvk the problem you're having is that hooks are not supported by `createRemoteComponent`. I'll update short-koala to use the non-hooks version

keithk commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`v0.19.0`](https://github.com/glitchdotcom/shared-components/releases/tag/v0.19.0) :rocket:
